Storage v2 auth schema fixes.

shamil-gadelshin opened the pull request:
This PR intends to update the authentication schema for the uploading. It changes the runtime, type, and storage-node projects.

#### Changes
- remove obsolete `authentication_key` from uploading parameters
- add `MaxDataObjectSize` runtime constant (defaulted to 10 GB)
- add `ipfs_content_id` to the `DataObject` runtime state
- add CID and size verification on the data uploading

shamil-gadelshin committed:
runtime: storage: Change data objects data.

- introduce MaxDataObjectSize constant
- set MaxDataObjectSize in runtime to 10 GB
- remove authentication_key from data uploading
- add ipfs_content_id to data object storage

shamil-gadelshin committed:
storage-node-v2: Improve auth schema.

- update types: add ipfsContentId to DataObject and remove authenticationKey
- add check for data object size, accepted status and multihash

shamil-gadelshin committed:
storage-node-v2: Fix build error.

Lezek123 commented on storage-node-v2/src/services/webApi/controllers/publicApi.ts:
```diff
@@ -107,10 +110,14 @@ export async function uploadFile(req: express.Request, res: express.Response): P
 
-    verifyFileSize(fileObj.size)
+    const api = getApi(res)
+    await verifyFileSize(api, fileObj.size)
```
Instead of checking the filesize after it's fully uploaded, perhaps you could use `limits.fileSize` property in `multer` config? (`OpenApiValidator.middleware` -> `fileUploader`)

shamil-gadelshin replied:
Very nice finding! Implemented.

Lezek123 commented on storage-node-v2/src/services/webApi/controllers/publicApi.ts:
```diff
@@ -296,17 +301,59 @@ async function validateTokenRequest(api: ApiPromise, tokenRequest: UploadTokenRe
+
+  if (dataObject.accepted.valueOf()) {
+    throw new WebApiError(`Data object had been already accepted ID = ${dataObjectId}`, 400)
+  }
```
I'm not sure what was the conclusion on allowing already accepted objects, since if the "accepting" node goes down before propagating the data - reuploading it will become impossible.

Another issue is that this opens up an attack vector where storage provider could effectively block lots of uploads by sending `acceptDataObject` transactions without ever recieving the objects, which may be an issue even considering that the provider will then be slashed at some point.

And the final note is that this check will not always work either, because some other `acceptDataObject` tx(s) may be already pending on-chain confirmation (or will be sent/recieved by other nodes earlier). Therefore my suggestion is to not use it for anything other than skipping an attempt to send `acceptDataObject` if the object has already been accepted (but still keeping the actual uploaded data in that case)

shamil-gadelshin replied:
We haven't decided on the 'allowing already accepted objects' issue yet, and I didn't add the 'data object accepted' check on the runtime side. I was planning to add the 'storage node side' check as a fail-fast solution. 
Regarding possible attack vectors from the storage providers - transaction fees and slashing are our current mitigations.

However, I like your idea because we are still thinking about re-uploading data objects on the failure of the single holder (we also need to think about the DoS from the client of the storage node - multiple uploading of the same large file if we allow re-uploading the data object).

Anyway, It's easy to make it strict again (possibly during the auth-token creation to counter mentioned DoS).

Lezek123 commented on storage-node-v2/src/services/webApi/controllers/publicApi.ts:
```diff
@@ -296,17 +301,59 @@ async function validateTokenRequest(api: ApiPromise, tokenRequest: UploadTokenRe
+
+  // Cannot get 'size' as a regular property.
+  const probablyDataObjectSize = dataObject.get('size') as unknown
+  const dataObjectSize = probablyDataObjectSize as BN
```
```suggestion
  const dataObjectSize = dataObject.getField('size')
```

shamil-gadelshin replied:
Fixed (I got errors previously using this code but I couldn't reproduce it again).

Lezek123 commented:
Tested the uploads, I noticed one issue when requesting the token with invalid signature:
```
Internal Server Error: no schema defined for status code '401' in the openapi spec
```

Other than that everything seems to work fine

shamil-gadelshin commented:
> Internal Server Error: no schema defined for status code '401' in the openapi spec

Fixed.

shamil-gadelshin committed:
storage-node-v2: Review fixes.

- added 401 status to the api spec (auth token endpoint)
- added filesize limit to the multer(Busboy) library settings
- changed ‘accepted data object’ check

Lezek123 commented on storage-node-v2/src/services/webApi/app.ts:
```diff
@@ -103,6 +114,8 @@ function validateUpload(api: ApiPromise, account: KeyringPair): ValidateUploadFu
   /* eslint-disable @typescript-eslint/no-unused-vars */
   return (req: express.Request, scopes: string[], schema: OpenAPIV3.SecuritySchemeObject) => {
+    return true
+
```
I think this got comitted by accident
```suggestion
```

shamil-gadelshin replied:
Yes. I implemented a flag disabling the auth in my latest branch (but not here).

Lezek123 reviewed (changes requested):
Did a few more tests, eveything works pretty much as expected, just noticed one more issue with the upload responses:

**Example 1:** (no space left)
```
{"type":"request_validation","message":"ENOSPC: no space left on device, write"}
```
**Example 2:** (uploads dir has been deleted)
```
{"type":"request_validation","message":"ENOENT: no such file or directory, open 'uploads-dir/d030312187891a709f787fde94bec1e8'"}
```

It seems like all errors that reach express error handler are treated as `request_validation` errors, regardless of the actual cause.

shamil-gadelshin committed:
Update storage-node-v2/src/services/webApi/app.ts

Co-authored-by: Lezek123 <leszek@jsgenesis.com>

shamil-gadelshin committed:
storage-node-v2: Change request error handling.

Lezek123 reviewed (approved):
LGTM!
